qiita_db/handlers/oauth2.py

```diff
@@ -0,0 +1,280 @@
+        user = db_token.get('user', None)
+        if user:
+            if not self.check_rate_limit(db_token['client_id'], user):
+                oauth_error(self, 'Oauth2 error: daily request limit reached',
+                            'invalid_grant')
+                return
+
+        return f(self, *args, **kwargs)
+    return wrapper
+
+
+class OauthBaseHandler(RequestHandler):
+    def write_error(self, status_code, **kwargs):
+        """Overriding the default write error in tornado RequestHandler"""
+        if status_code in {403, 404, 405}:
+            # We don't need to log these failues in the logging table
+            return
+        # log the error
+        from traceback import format_exception
+        exc_info = kwargs['exc_info']
+        trace_info = ''.join(['%s\n' % line for line in
+                             format_exception(*exc_info)])
+        req_dict = self.request.__dict__
+        # must trim body to 1024 chars to prevent huge error messages
+        req_dict['body'] = req_dict.get('body', '')[:1024]
+        request_info = ''.join(['<strong>%s</strong>: %s\n' %
+                               (k, req_dict[k]) for k in
+                                req_dict.keys() if k != 'files'])
+        error = exc_info[1]
+        qdb.logger.LogEntry.create(
+            'Runtime',
+            'ERROR:\n%s\nTRACE:\n%s\nHTTP INFO:\n%s\n' %
+            (error, trace_info, request_info))
+
+    def head(self):
+        """Adds proper response for head requests"""
+        self.finish()
+
+    def oauth_error(self, error_msg, error='invalid_request'):
+        self.set_status(400)
+        self.write({'error': error,
+                    'error_description': error_msg})
+        self.finish()
+
+    def check_rate_limit(self, client_id, user):
+        limit_key = '%s_%s_daily_limit' % (client_id, user)
+        limiter = r_client.get(limit_key)
+        if limiter is None:
+            # Set limit to 5,000 requests per day
+            r_client.setex(limit_key, 5000, 86400)
+        else:
+            r_client.decr(limit_key)
+            if int(r_client.get(limit_key)) <= 0:
+                return False
+        return True
+
+
+class TokenAuthHandler(OauthBaseHandler):
+    def generate_access_token(self):
+        """Creates the random alphanumeric token
+
+        Returns
+        -------
+        str
+            55 character random alphanumeric string
+        """
+        pool = ascii_letters + digits
+        return ''.join((SystemRandom().choice(pool) for _ in range(55)))
+
+    def set_token(self, token, client_id, user=None, timeout=3600):
+        """Create the access token for the client on redis
+
+        Parameters
+        ----------
+        token: str
+            Random token string for authorization
+        client_id : str
+            Client that requested the token
+        user : str, optional
+            If password grant type requested, the user requesting the key.
+        timeout : int
```

Comment:
Do you have links that support that 1 day is too long for registration token? I would like to have at lest more than a single our, to avoid the plugins requesting it multiple times... Most of our jobs are longer than an hour.